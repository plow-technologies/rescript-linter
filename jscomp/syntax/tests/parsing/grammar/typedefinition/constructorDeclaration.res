type t = Rgb
type t =
  | Rgb

type t =
  | White(grayscale)
  | Black(grayscale,) // trailing comma
  | Rgb(int, int, int) 
  | Rgba(int, int, int, int,) // trailing comma

type t =
  | Rgb({r: int, g: int, b: int})
  // trailing comma
  | Rgba({r: int, g: int, b: int, a: int,},)
  | JsColor({"gradient": int})
  | JsColor({"gradient": int,})
  | JsColor({"gradient": int},)
  | JsColor({@attr "gradient": int,},)
  | JsColor({"gradient": int}, color)
  | JsColor({"gradient": int}, color,)
  | JsColor({"gradient": int}, {"hex": string}, int)
  | JsColor({@attr "gradient": int}, {@attr "hex": string}, int)
  | JsT({..}) // Note: this compiles to bucklescript
  | JsT({.. "gradient": int}) // Note: this compiles to bucklescript
  | JsT({.. "gradient": int,}) // Note: this compiles to bucklescript
  | JsT({.. @attr "gradient": int,}) // Note: this compiles to bucklescript
  | JsT({.. "gradient": int, "hex": string}, {.. "gradient": int, "hex": string}) // Note: this compiles to bucklescript
  | JsT(
      {.. @attr "gradient": int, @attr "hex": string,},
      {.. @attr "gradient": int, @attr "hex": string,},
    ) // Note: this compiles to bucklescript

// gadt
type t = Rgb : t
type t =
  | Rgb : t

type t =
  | White(grayscale) : t
  | Black(grayscale,) : t// trailing comma
  | Rgb(int, int, int) : t
  | Rgba(int, int, int, int,) : t // trailing comma

type t =
  | Rgb({r: int, g: int, b: int}) : t
  // trailing comma
  | Rgba({r: int, g: int, b: int, a: int,},) : t
  | JsColor({"gradient": int}) : t
  | JsColor({"gradient": int,}) : t
  | JsColor({"gradient": int},) : t
  | JsColor({"gradient": int,},) : t
  | JsColor({@attr "gradient": int,},) : t
  | JsColor({"gradient": int}, color) : t
  | JsColor({"gradient": int}, color,) : t
  | JsColor({"gradient": int}, {"hex": string}, int) : t
  | JsT({..}) : t // Note: this compiles to bucklescript
  | JsT({.. "gradient": int}) : t // Note: this compiles to bucklescript
  | JsT({.. @attr "gradient": int}) : t // Note: this compiles to bucklescript
  | JsT({.. "gradient": int,}) : t // Note: this compiles to bucklescript
  | JsT({.. @attr "gradient": int,}) : t // Note: this compiles to bucklescript
  | JsT({.. "gradient": int,},) : t // Note: this compiles to bucklescript
  | JsT(
      {.. "gradient": int, "hex": string},
      {.. "gradient": int, "hex": string}
    ) : t // Note: this compiles to bucklescript
  | JsT(
      {.. @attr "gradient": int, @attr "hex": string,},
      {.. @attr "gradient": int, @attr "hex": string,},
    ) : t // Note: this compiles to bucklescript

// attributes
// type t = @attr EmptyColor
type t = 
  | @attr EmptyColor
  | @onConstr White(grayscale) : @onGadt t

type node <_, 'value> =
  | Root({
      mutable value: 'value,
      mutable updatedTime: float,
    }): node<root, 'value>
  | Derived({
      mutable cachedValue: 'value,
      parent: node<_, 'value>,
      root: node<root, 'value>,
      updateF: 'value => 'value,
      mutable updatedTime: float,
    }): node<derived, 'value>

type delta = Compute({"blocked_ids": unit} => unit)
type queryDelta =
  | Compute({"blocked_ids": unit} => unit)
  | Compute({"blocked_ids": unit} => unit, {"allowed_ids": unit} => unit)
