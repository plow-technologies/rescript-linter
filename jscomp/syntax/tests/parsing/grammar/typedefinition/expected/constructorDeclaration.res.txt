type nonrec t =
  | Rgb 
type nonrec t =
  | Rgb 
type nonrec t =
  | White of grayscale 
  | Black of grayscale 
  | Rgb of int * int * int 
  | Rgba of int * int * int * int 
type nonrec t =
  | Rgb of {
  r: int ;
  g: int ;
  b: int } 
  | Rgba of {
  r: int ;
  g: int ;
  b: int ;
  a: int } 
  | JsColor of < gradient: int   >  
  | JsColor of < gradient: int   >  
  | JsColor of < gradient: int   >  
  | JsColor of < gradient: int [@attr ]  >  
  | JsColor of < gradient: int   >  * color 
  | JsColor of < gradient: int   >  * color 
  | JsColor of < gradient: int   >  * < hex: string   >  * int 
  | JsColor of < gradient: int [@attr ]  >  * < hex: string [@attr ]  >  *
  int 
  | JsT of < .. >  
  | JsT of < gradient: int   ;.. >  
  | JsT of < gradient: int   ;.. >  
  | JsT of < gradient: int [@attr ]  ;.. >  
  | JsT of < gradient: int  ;hex: string   ;.. >  *
  < gradient: int  ;hex: string   ;.. >  
  | JsT of < gradient: int [@attr ] ;hex: string [@attr ]  ;.. >  *
  < gradient: int [@attr ] ;hex: string [@attr ]  ;.. >  
type nonrec t =
  | Rgb: t 
type nonrec t =
  | Rgb: t 
type nonrec t =
  | White: grayscale -> t 
  | Black: grayscale -> t 
  | Rgb: int * int * int -> t 
  | Rgba: int * int * int * int -> t 
type nonrec t =
  | Rgb: {
  r: int ;
  g: int ;
  b: int } -> t 
  | Rgba: {
  r: int ;
  g: int ;
  b: int ;
  a: int } -> t 
  | JsColor: < gradient: int   >  -> t 
  | JsColor: < gradient: int   >  -> t 
  | JsColor: < gradient: int   >  -> t 
  | JsColor: < gradient: int   >  -> t 
  | JsColor: < gradient: int [@attr ]  >  -> t 
  | JsColor: < gradient: int   >  * color -> t 
  | JsColor: < gradient: int   >  * color -> t 
  | JsColor: < gradient: int   >  * < hex: string   >  * int -> t 
  | JsT: < .. >  -> t 
  | JsT: < gradient: int   ;.. >  -> t 
  | JsT: < gradient: int [@attr ]  ;.. >  -> t 
  | JsT: < gradient: int   ;.. >  -> t 
  | JsT: < gradient: int [@attr ]  ;.. >  -> t 
  | JsT: < gradient: int   ;.. >  -> t 
  | JsT: < gradient: int  ;hex: string   ;.. >  *
  < gradient: int  ;hex: string   ;.. >  -> t 
  | JsT: < gradient: int [@attr ] ;hex: string [@attr ]  ;.. >  *
  < gradient: int [@attr ] ;hex: string [@attr ]  ;.. >  -> t 
type nonrec t =
  | EmptyColor [@attr ]
  | White: grayscale -> ((t)[@onGadt ]) [@onConstr ]
type nonrec (_, 'value) node =
  | Root: {
  mutable value: 'value ;
  mutable updatedTime: float } -> (root, 'value) node 
  | Derived:
  {
  mutable cachedValue: 'value ;
  parent: (_, 'value) node ;
  root: (root, 'value) node ;
  updateF: 'value -> 'value ;
  mutable updatedTime: float } -> (derived, 'value) node 
type nonrec delta =
  | Compute of (< blocked_ids: unit   >  -> unit) 
type nonrec queryDelta =
  | Compute of (< blocked_ids: unit   >  -> unit) 
  | Compute of (< blocked_ids: unit   >  -> unit) *
  (< allowed_ids: unit   >  -> unit) 